keps/sig-scheduling/2044-topology-awareness-in-kube-scheduler/README.md

```diff
@@ -0,0 +1,767 @@
+
+	TopologyPolicies []string 		   `json:"topologyPolicies"`
+	Cells            map[string]Cell   `json:"cells"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// NodeResourceTopologyList is a list of NodeResourceTopology resources
+type NodeResourceTopologyList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+
+	Items []NodeResourceTopology `json:"items"`
+}
+
+// Cell is the spec for a NodeResourceTopology resource
+type Cell struct {
+    Name       string           `json:"name"`
+    Type       string           `json:"type"`
+    Parent     string           `json:"parent,omitempty"`
+    Costs      []CostInfo       `json:"costs,omitempty"`
+    Attributes []AttributeInfo  `json:"attributes,omitempty"`
+    Resources  []ResourceInfo   `json:"resources,omitempty"`
+}
+
+type ResourceInfo struct {
+    Name        string `json:"name"`
+    Allocatable string `json:"allocatable"`
+    Capacity    string `json:"capacity"`
+}
+
+type CostInfo struct {
+    Name  string `json:"name"`
+    Value int	 `json:"value"`
+}
+
+type AttributeInfo struct {
+    Name  string `json:"name"`
+    Value string `json:"value"`
+}
+// Kubelet writes to NodeResourceTopology
+// and scheduler plugin reads from it 
+// Real world example of how these fields are populated is as follows:
+// Cells:
+//   Name:     node-1
+//   Type:     Node
+//   Costs:
+//     Name:   node-0
+//     Value:  20
+//     Name:   node-1
+//     Value:  10
+//   Attributes:
+//    Name: performance-profile
+//    Value: high-performance-profile
+//   Resources:
+//     Name:         example.com/deviceB
+//     Allocatable:  2
+//     Capacity:     2
+//     Name:         example.com/deviceA
+//     Allocatable:  2
+//     Capacity:     2
+//     Name:         cpu
+//     Allocatable:  4
+//     Capacity:     4
+
+```
+
+
+### Plugin implementation details
+
+### Description of the Algorithm
+
+The algorithm which of the scheduler plugin is as follows:
+
+1. At the filter extension point of the plugin, the QoS class of the pod is determined, in case it is a best effort pod or the 
+   Topology Manager Policy configured on the node is not single-numa-node, the node is not considered for scheduling
+1. The Topology Manager Scope is determined.
+1. While interating through the containers of a pod
+	* A bitmask is created where each bit corresponds to a NUMA cell and are all the bits are set. If the resources cannot be aligned on the NUMA cell,
+	the bit should be unset.
+	* For each resource requested in a container, a new resourceBitmask is created to determined which NUMA cell is a good fit for each resource
+		1. If requested resource cannot be found on a node, it is unset as available NUMA cell
+		1. If an unknown resource has 0 quantity, the NUMA cell should be left set.
+	* The following checks are performed:
+		1. Add NUMA cell to the resourceBitmask if resource is cpu and it's not guaranteed QoS, since cpu will flow
```

Comment:
Do constants within the API for 'cpu' and 'memory' make sense?

Reply:
... perhaps others?

Reply:
"cpu", "memory" and other resources are not stored as constants but rather stored in the ResourceInfo data structure below where the name corresponds the resource name stored here as value which is obtained from the v1.ResourceCPU or  v1.ResourceMemory and in the scheduler plugin a resource is compared with values v1.ResourceCPU or  v1.ResourceMemory to determine if the resource is a CPU or memory.

```
type ResourceInfo struct {
    Name        string `json:"name"`
    Allocatable string `json:"allocatable"`
    Capacity    string `json:"capacity"`
}
```